tensorflow_lite_support/cc/task/processor/image_postprocessor.cc

```diff
@@ -0,0 +1,237 @@
+ImagePostprocessor::Create(core::TfLiteEngine* engine,
+                           const std::initializer_list<int> output_indices,
+                           const std::initializer_list<int> input_indices) {
+  ASSIGN_OR_RETURN(auto processor,
+                   Processor::Create<ImagePostprocessor>(
+                       /* num_expected_tensors = */ 1, engine, output_indices,
+                       /* requires_metadata = */ false));
+
+  RETURN_IF_ERROR(processor->Init(input_indices));
+  return processor;
+}
+
+absl::Status ImagePostprocessor::Init(const std::vector<int> &input_indices) {
+  if (core::TfLiteEngine::OutputCount(engine_->interpreter()) != 1) {
+    return tflite::support::CreateStatusWithPayload(
+        absl::StatusCode::kInvalidArgument,
+        absl::StrFormat(
+            "Image segmentation models are expected to have only 1 "
+            "output, found %d",
+            core::TfLiteEngine::OutputCount(engine_->interpreter())),
+        tflite::support::TfLiteSupportStatus::kInvalidNumOutputTensorsError);
+  }
+
+  if (GetTensor()->type != kTfLiteUInt8 &&
+      GetTensor()->type != kTfLiteFloat32) {
+    return tflite::support::CreateStatusWithPayload(
+        absl::StatusCode::kInvalidArgument,
+        absl::StrFormat("Type mismatch for output tensor %s. Requested one "
+                        "of these types: "
+                        "kTfLiteUint8/kTfLiteFloat32, got %s.",
+                        GetTensor()->name,
+                        TfLiteTypeGetName(GetTensor()->type)),
+        tflite::support::TfLiteSupportStatus::kInvalidOutputTensorTypeError);
+  }
+
+  if (GetTensor()->dims->data[0] != 1 || GetTensor()->dims->data[3] != 3) {
+    return CreateStatusWithPayload(
+        absl::StatusCode::kInvalidArgument,
+        absl::StrCat("The input tensor should have dimensions 1 x height x "
+                     "width x 3. Got ",
+                     GetTensor()->dims->data[0], " x ",
+                     GetTensor()->dims->data[1], " x ",
+                     GetTensor()->dims->data[2], " x ",
+                     GetTensor()->dims->data[3], "."),
+        tflite::support::TfLiteSupportStatus::
+            kInvalidInputTensorDimensionsError);
+  }
+
+  // Gather metadata
+  const tflite::TensorMetadata* output_metadata =
+      engine_->metadata_extractor()->GetOutputTensorMetadata(
+          tensor_indices_.at(0));
+  const tflite::TensorMetadata* input_metadata =
+      engine_->metadata_extractor()->GetInputTensorMetadata(
+          input_indices.at(0));
+
+  // Use input metadata for normalization as fallback.
+  const tflite::TensorMetadata* processing_metadata =
+      GetNormalizationOptionsIfAny(*output_metadata).value().has_value()
+          ? output_metadata
+          : input_metadata;
+
+  absl::optional<vision::NormalizationOptions> normalization_options;
+  ASSIGN_OR_RETURN(normalization_options,
+                   GetNormalizationOptionsIfAny(*processing_metadata));
```

Comment:
Simplify this by something like:
```
ASSIGN_OR_RETURN(auto normalization_options, GetNormalizationOptionsIfAny(*GetTensorMetadata()));

if (!normalization_options.has_value() && input_index>-1) {
  ASSIGN_OR_RETURN(normalization_options, GetNormalizationOptionsIfAny(*input_metadata));
}
```

Reply:
Why re-assign twice? I suppose you meant

```c++
  ASSIGN_OR_RETURN(auto normalization_options, GetNormalizationOptionsIfAny(*GetTensorMetadata()));

  if (GetTensor()->type == kTfLiteFloat32) {
    if (!normalization_options.has_value() && input_index > -1) {
     ....
```
No?

Reply:
Nice one! 👍